### ISSUE TITLE
How is CycleGAN training generators?

### ISSUE THREAD

**Issue by HughRunyan** [junyanz/pytorch-CycleGAN-and-pix2pix#717]
I don't see any gradient reversal layers or freezing the discriminators to only train generators. Is cycle consistency loss enough to train the generators? Shouldn't there be a GAN loss training phase for the generators? Am I missing something in the code? Thanks for your time and any help

**Comment by junyanz** [comment]
We calculate GAN training loss in this [line](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/cycle_gan_model.py#L169).

**Comment by HughRunyan** [comment]
Thanks! That line calls the GANLoss module over in networks.py. I'm assuming you're using MSE loss. My fundamental question is, are you training the generator with the correct ground truth labels? To me this would mean you're training the discriminator correctly, but giving the generator the wrong gradient, because we want to optimize it to trick the discriminator. If you're training with incorrect ground truth labels, doesn't this mess up the discriminator at the expense of passing the desired signal to the generator? Is it just necessary to train the discriminator incorrectly in order to train the generator, and this is overcome by later detaching the generator to train just the discriminator with the correct ground truths?

**Comment by HughRunyan** [comment]
That wasn't super clear, I'm sorry...the point I'm trying to get at is, how specifically are you training the generators through the discriminators?

**Comment by junyanz** [comment]
I recommend that you read the [DCGAN](https://github.com/pytorch/examples/blob/master/dcgan/main.py) code and the original GAN paper. The objective for generator and discriminator is slightly different, and our code reflects the difference.
